drivers/pci/controller/vmd.c

```diff
@@ -767,17 +814,253 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
+
+	vmd->resources[2] = (struct resource){
+		.name = "VMD MEMBAR2",
+		.start = res->start + membar2_offset,
+		.end = res->end,
+		.flags = flags,
+		.parent = res,
+	};
+
+	if (vmd_has_pch_rootbus(vmd)) {
+		pci_read_config_dword(vmd->dev, PCI_MEMBAR2_OFFSET1, &reg);
+		vmd_enh_membar2_offset = reg;
+		pci_read_config_dword(vmd->dev, PCI_MEMBAR2_OFFSET2, &reg);
+		vmd_enh_membar2_offset |= (u64)reg << 32;
+
+		vmd->resources[5] = (struct resource){
+			.name = "VMD MEMBAR2 PCH",
+			.start = res->start + membar2_offset +
+				 vmd_enh_membar2_offset,
+			.end = res->end,
+			.flags = flags,
+			.parent = &vmd->resources[2],
+		};
+
+		/*
+		 * Update first MEMBAR2 range (IOC), memory start does not
+		 * change, memory end is one less than the offset stored
+		 * as a 64 bit value in PCI_MEMBAR2_OFFSET1 and
+		 * PCI_MEMBAR2_OFFSET2 registers.
+		 */
+		vmd->resources[2].end = res->start + vmd_enh_membar2_offset +
+				       membar2_offset - 1;
+	}
+}
+
+static void vmd_bus_enumeration(struct pci_bus *bus, unsigned long features)
+{
+	struct pci_bus *child;
+	struct pci_dev *dev;
+	int ret;
+
+	vmd_acpi_begin();
+
+	pci_scan_child_bus(bus);
+	vmd_domain_reset(vmd_from_bus(bus));
+
+	/*
+	 * When Intel VMD is enabled, the OS does not discover the Root Ports
+	 * owned by Intel VMD within the MMCFG space. pci_reset_bus() applies
+	 * a reset to the parent of the PCI device supplied as argument. This
+	 * is why we pass a child device, so the reset can be triggered at
+	 * the Intel bridge level and propagated to all the children in the
+	 * hierarchy.
+	 */
+	list_for_each_entry(child, &bus->children, node) {
+		if (!list_empty(&child->devices)) {
+			dev = list_first_entry(&child->devices, struct pci_dev,
+					       bus_list);
+			ret = pci_reset_bus(dev);
+			if (ret)
+				pci_warn(dev, "can't reset device: %d\n", ret);
+
+			break;
+		}
+	}
+
+	pci_assign_unassigned_bus_resources(bus);
+
+	pci_walk_bus(bus, vmd_pm_enable_quirk, &features);
+
+	/*
+	 * VMD root buses are virtual and don't return true on pci_is_pcie()
+	 * and will fail pcie_bus_configure_settings() early. It can instead be
+	 * run on each of the real root ports.
+	 */
+	list_for_each_entry(child, &bus->children, node)
+		pcie_bus_configure_settings(child);
+
+	pci_bus_add_devices(bus);
+
+	vmd_acpi_end();
+}
+
+static int vmd_create_pch_bus(struct vmd_dev *vmd, struct pci_sysdata *sd,
+			      resource_size_t *offset)
+{
+	LIST_HEAD(resources_pch);
+
+	pci_add_resource(&resources_pch, &vmd->resources[3]);
+	pci_add_resource_offset(&resources_pch, &vmd->resources[4],
+				offset[0]);
+	pci_add_resource_offset(&resources_pch, &vmd->resources[5],
+				offset[1]);
+
+	vmd->bus_pch =
+		pci_create_root_bus(&vmd->dev->dev, vmd->busn_start_pch,
+					&vmd_ops, sd, &resources_pch);
+
+	if (!vmd->bus_pch) {
+		pci_free_resource_list(&resources_pch);
+		vmd_remove_irq_domain(vmd);
+		return -ENODEV;
+	}
+
+	/*
+	 * VMD enhacement specific: pci_scan_bridge_extend() in probe.c
```

Comment:
skróciłbym to trochę, i na razie podkreślił że to jest te miejsce przez które robimy RFC
```
This is a workaround for pci_scan_bridge_extend() code.
It assigns setup as broken when primary != bus->number.
For PCH rootbus primary is not "hard-wired to 0".
To avoid this, vmd->bus_pch->number and vmd->bus_pch->primary  are updated to same value.
```

Reply:
Fixed.